### ISSUE TITLE
Shift-click to select multiple cells in Units Table

### ISSUE THREAD

**Issue by khl02007** [magland/sortingview#92]
Sometimes the user sorts the cells based on a column (e.g. noise overlap) in the Units Table. The user may then want to select all cells below certain threshold value and give them a particular label. Right now there is no easy way to select multiple cells at once. Would be nice if shift-click selects a range of cells.

**Comment by jsoules** [comment]
To be clear on the functionality request here, I am interpreting this as "I would like to select one checkbox, then shift-click another checkbox, and have it check every checkbox in a row that's currently laid out between the two clicks."

**Comment by khl02007** [comment]
@magland @jsoules Yes I think I was looking for just 

> To be clear on the functionality request here, I am interpreting this as "I would like to select one checkbox, then shift-click another checkbox, and have it check every checkbox in a row that's currently laid out between the two clicks."

but maybe that's not specific enough to handle all cases you're thinking of? I agree with Jeremy that ctrl-click for inverting the current checkbox list is unintuitive. As for the shift-click, I think that shift-clicking on box _t_ should check every box between _t_ and an upstream box _k_ that's already been checked. If _k_ doesn't exist (i.e. no box is checked upstream of _t_) then shift-click probably should behave like normal click. Basically I was thinking of something like what happens when you shift-click in a file manager.

**Comment by jsoules** [comment]
I was thinking the ctrl-click invert-selection feature would let you select a large but discontinuous number of items easily; but if that's not something that comes up much, I'll disable it. (I suppose you can always get the same effect by doing a select-all and then clicking the ones you want to deselect.)

For the shift-click, the reason I switched from my first proposal to what I eventually implemented was that this version seemed clearer & more expressive:
- It lets you set bounds on both sides for the range of boxes to be checked (so your selected range can be both upstream or downstream)
- You can always see from the list exactly which items will wind up selected (the new selection range never crosses a boundary that contains an already-selected box)
- It works using the list state only, so we don't need to begin tracking the most-recently-clicked checkbox
- It avoids potentially tricky situations where it's unclear what the 'prior-click' box is:
  - If you have several boxes checked, then you re-sort the list, it can be tough to keep track of which box is the anchor
  - Does the anchor reset when you click the table header to re-sort the list? I think we could implement this reasonably, but it'd be a little confusing that that interaction changes the anchor but other non-checkbox clicks would not.
  - Does the anchor reset if you click in a different widget? The only way to do this I can think of (since one widget can't listen to another's events) is really gnarly, we'd have to implement a universal onClick handler that causes the MountainView component, if any, to reset the anchor checkbox.
  - Does the anchor reset if the most recent click was to uncheck a box? If so, what's the new anchor?

Plus if there's only one box checked, it just defaults to the behavior you first described: check the upstream box, then shift-click the box below it, and everything below the checked box gets checked.

That said, if you're still unconvinced, I'll also try to implement it as originally described.

**Comment by khl02007** [comment]
@jsoules I see, in that case feel free to implement the version you think is best -  I don't have a strong opinion.

**Comment by jsoules** [comment]
@khl02007 Hi Kyu, thanks--I would like for you to get the chance to play around with it before we make a final decision though--I don't want to bully you into features that aren't fully meeting your needs!

I've removed the ctrl-click functionality from the current version of this branch on Github; let me know if you'd like to download it and check it out by running a local backend (please email me directly if what's in this readme isn't clear enough). Or we can sort out how we can host a version for you to look at.

**Comment by khl02007** [comment]
No worries - I agree that it may be useful us (not just me but others in the lab) to try out the feature on real recordings (esp with a lot of cells like Alison's data set). Is there a way we can do that?

**Comment by jsoules** [comment]
Hi Kyu,
Sorry it's taken me a bit to figure out the best way to do this--but I have a test link for you now!

You can try this:
https://sortingview-jsoules-spikeforest.vercel.app/workspace/sorting/R-c5089c48b72f/S-816a8e175985?workspace=e5c306c436a9f75fd97a20d832cd54d8243d516da53cbfe86270f11e98ee8b26&channel=franklab

This is connected to the same data set that we were using to explore slowness issues in #105--I actually just finished porting over the changes in this PR so that they are compatible with the changes that improved performance in that issue. If you want to experiment with other live data sets, you can use the production version to open whatever data set you want to see. Copy the `/workspace/sorting/...` part and replace the `/workspace/sorting/...` part in the link above, and that should open the corresponding workspace for you. We'll try to make this available for you for any future fixes as well.

NOTE: this is connected to the **live data**, not a test copy. Any curation changes you make will actually be applied to the live data set!

I think the best way to describe the behavior I implemented here is that it works like a paint bucket tool in a graphics editor. If you shift-click, it will 'fill' a continuous range with checked/unchecked boxes, up until it hits a boundary (either a box with the opposite state, or the edge of the list). So in this data set, if you want to select all units with more than 4000 spikes, you would:

- Open the Units Table view
- Click 'show all rows' at the bottom (to interact with the full data set)
- Click to sort by num_spikes
- Scroll down to the first row that has >4000 spikes (should be unit 8)
- Click to select unit 8
- Then shift-click any checkbox below unit 8 to select all units between that one and the end of the list

I think this is pretty much the workflow you had in mind, but it also supports toggling a region that doesn't go to the end--like in this example, if you followed the steps above, then click (no shift) to deselect units 213 and 84, and then shift-click the checkbox for any of the selected rows in between them, you'll deselect that block of rows. (IDs 63, 26, 191, 101, and 220)

Anyway, let me know what you think! If this meets your needs we'll get it merged.

**Comment by acomrie** [comment]
Hi all, I tried this out today and also asked the two others I'm working with to try it too. My overall sense is that although the new functionality may be useful, it is not as intuitive as it could be for users.

The shift+click functionality as it stands does work in testing, and is especially helpful for thresholding-type situations as Jeff suggested. However, people have pretty strong priors about what "click; shift+click" will do in this kind of row-based situation (as in any file browser GUI). So, I'd propose giving shift+click a more standard functionality, such that it only selects units displayed between and including two particular units (as Jeff said, "'I would like to select one checkbox, then shift-click another checkbox, and have it check every checkbox in a row that's currently laid out between the two clicks'"). This way shift+click will simply do what people expect most intuitively.

I think the other functionality you've implemented could be handy to associate with some other key combination rather than shift+click, but is lower priority in my opinion. If this gets implemented, it would be helpful to have a little documentation/instructions note somewhere obvious for users to learn how to use it, like a little "tips" button or section in one of the menus.

I see the ctrl+click functionality (to deselect already-selected units) similarly; could be useful (especially if it is implemented in the way that users will already be familiar with from general computer/file browser gui use), but isn't as important as getting shift+click merged.

Thanks for your work on this - having the test link was super helpful!

**Comment by jsoules** [comment]
@acomrie Thanks for checking this out & for your feedback. It sounds like the paint-bucket approach isn't intuitive, so I'll implement something else (& probably keep that around--maybe as ctrl-click (?) since the standard behavior for that interaction--toggling selection of one list item without changing the others--is what you already get from a checkbox list).

To make sure I implement this right, I'd like to ask the team to collectively let me know what you would expect to happen in the following scenarios:

Suppose we have a continuous list of units sorted by unit ID. The user clicks unit 10, then shift-clicks unit 20. (I'm going to call clicking 10 and then shift-clicking 20, together, the 'interaction').

- If both 10 and 20 were unselected before the interaction, and everything between was also unselected, we'd expect the range of units [10, 20] (inclusive) to wind up selected.
- If both 10 and 20 were checked before the interaction (so that clicking 10 de-selected it), and everything in between was also selected, we'd expect the range of units [10, 20] to wind up unselected. Agree? Or should this *only* function for selection and not de-selection?
- If 10 was deselected before the interaction but 20 was selected before the interaction, we'd expect the range [10, 20] to wind up selected?
- Same situation reversed--10 selected before the interaction but 20 deselected--we'd expect the range to wind up de-selected? i.e. is the rule that the shift-click identifies a range with one end at the last click and one end at the unit shift-clicked, and applies over that range the same action that was applied by the click?
- Suppose 10 and 20 are both unchecked before the interaction, but the range [13, 15] is already checked. I'm expecting we don't care about the status of the intervening set, we just select/check everything in between? i.e. again, the action applied by the click in the interaction is what governs, and the shift just identifies a range for it?
- Does changing the sort order reset the last-clicked-unit? i.e. if I click 10, then re-sort, then shift-click, what (if anything) should happen?

**Comment by acomrie** [comment]
> If both 10 and 20 were unselected before the interaction, and everything between was also unselected, we'd expect the range of units [10, 20] (inclusive) to wind up selected.

Agreed

> If both 10 and 20 were checked before the interaction (so that clicking 10 de-selected it), and everything in between was also selected, we'd expect the range of units [10, 20] to wind up unselected. Agree? Or should this only function for selection and not de-selection?

My first instinct is that the first click on 10 should deselect 10 and the shift-click on 20 should just act like a normal click only, deselecting 20. (10,20) would remain selected. Or perhaps shift+click on 20 would leave 20 selected because "click, shift+click" behavior should perhaps only have the opportunity to occur if the initial click is on an initially deselected unit. I don't feel too strongly about this though. A deselect functionality as you described it could be handy but in hopes of making this all very intuitive for folks, maybe we can wait on it? 

> If 10 was deselected before the interaction but 20 was selected before the interaction, we'd expect the range [10, 20] to wind up selected?

Agreed

> Same situation reversed--10 selected before the interaction but 20 deselected--we'd expect the range to wind up de-selected? i.e. is the rule that the shift-click identifies a range with one end at the last click and one end at the unit shift-clicked, and applies over that range the same action that was applied by the click?

My not-especially-strong instinct here again is that the interaction should only select; so, if 10 started as selected, then the first click would deselect 10 and the subsequent shift+click on 20 would either do nothing (or possibly act like a normal click).

> Suppose 10 and 20 are both unchecked before the interaction, but the range [13, 15] is already checked. I'm expecting we don't care about the status of the intervening set, we just select/check everything in between? i.e. again, the action applied by the click in the interaction is what governs, and the shift just identifies a range for it?

Agreed, if [13,15] are preselected, but 10 is not, and 10 is clicked, then 20 is shift-clicked, then [10,20] would become selected. If [13,15] are preselected and 10 is also selected, and 20 is deselected, then clicking on 20 followed by shift+click on 10 would select [10,20] I think, too. Having the shift click functionality work in both ascending and descending orders may also help get around some of the previously raised situations.

> Does changing the sort order reset the last-clicked-unit? i.e. if I click 10, then re-sort, then shift-click, what (if anything) should happen?

Don't think it needs to reset... I think the first click on 10 would select that unit, then resort occurs, then shift click on 20 regardless of its selection status would select [10,20] where the units between 10 and 20 are from the most recent (not the initial) sorted/ordered display of units. Though I don't expect this to be a super common situation, so probably doesn't matter too much how this works.

> I'm assuming that this entire set of interactions is not expected to change the selection status of anything outside the range from click to shift-click. So like, where in a file browser if you had selected files Q, A, and D by ctrl-clicks (in that order), then shift-clicked G, you'd wind up with ONLY D, E, F, G selected--but for this feature we probably don't want that?

I think I agree here. So instead of the file browser behavior, seems like here we'd want the shift+click of G to result in Q, A, and D, E, F, G to be selected. This would enable users to additively select several ranges of units with multiple subsequent "click, shift+click" interactions.

Hope that helps! Certainly open to opinions and ideas from other folks.

**Comment by khl02007** [comment]
@jsoules sorry for being late to this discussion. I actually like the way you have implemented it (and the paint bucket analogy is apt indeed!). What I especially like is that it makes it easy to select a lot of cells above/below a certain cell without having to scroll up/down. Honestly I could see myself using it, but what @acomrie describes also makes sense. I guess what I'm trying to say is...I don't have a strong opinion? I'm willing to adapt to whichever method.

Here are my answers to the questions that you asked, going by what I had in mind originally (may or may not be the same as what @acomrie said):

> Suppose we have a continuous list of units sorted by unit ID. The user clicks unit 10, then shift-clicks unit 20. (I'm going to call clicking 10 and then shift-clicking 20, together, the 'interaction').
> 
> * If both 10 and 20 were unselected before the interaction, and everything between was also unselected, we'd expect the range of units [10, 20] (inclusive) to wind up selected.

Yes
> * If both 10 and 20 were checked before the interaction (so that clicking 10 de-selected it), and everything in between was also selected, we'd expect the range of units [10, 20] to wind up unselected. Agree? Or should this _only_ function for selection and not de-selection?

Yes

> * If 10 was deselected before the interaction but 20 was selected before the interaction, we'd expect the range [10, 20] to wind up selected?

yes

> * Same situation reversed--10 selected before the interaction but 20 deselected--we'd expect the range to wind up de-selected? i.e. is the rule that the shift-click identifies a range with one end at the last click and one end at the unit shift-clicked, and applies over that range the same action that was applied by the click?

yes

> * Suppose 10 and 20 are both unchecked before the interaction, but the range [13, 15] is already checked. I'm expecting we don't care about the status of the intervening set, we just select/check everything in between? i.e. again, the action applied by the click in the interaction is what governs, and the shift just identifies a range for it?

yes

> * Does changing the sort order reset the last-clicked-unit? i.e. if I click 10, then re-sort, then shift-click, what (if anything) should happen?

not sure what this is asking. seems like it would depend on where you shift click?

> * I'm assuming that this entire set of interactions is not expected to change the selection status of anything outside the range from click to shift-click. So like, where in a file browser if you had selected files Q, A, and D by ctrl-clicks (in that order), then shift-clicked G, you'd wind up with ONLY D, E, F, G selected--but for this feature we probably don't want that?

yes

**Comment by acomrie** [comment]
I agree that the "paint bucket" description really helped me understand what you were going for and how to use it - maybe this suggests that keeping this functionality but with an alternative key combination would be good? So that shift click does what people intuitively expect shift click to do, and then the paint bucket approach is still available for folks to use too?

**Comment by khl02007** [comment]
@jsoules By the way, I just realized that Gmail has the exact shift click feature that I had in mind originally

**Comment by magland** [comment]
@jsoules what's the status of this?

**Comment by jsoules** [comment]
Fell off my radar while I was working on other things. I'll dig back into it.

**Comment by jsoules** [comment]
Hi all,

This request has been fulfilled with the functionality described by @khl02007 and @acomrie in September 2021, by Spikesortingview PR #41, and should now be live. Let me know if you encounter any issues.
